### PR TITLE
[SU-535] Generate "CANCELACIÓN DEFINITIVA POR REESTRUCTURACIÓN" Event…

### PR DESCRIPTION
**Pull request by oluexpert99**
… on Top-Up Loan Closure

## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 1]
```diff
@@ -1375,26 +1375,51 @@ private void createAnulacionNoveltyNews(Loan loan, LocalDate writeOffDate) {
...
-            if (loanCharge.getAmountOutstanding(loan.getCurrency()).isGreaterThanZero()) {
-                if ((incident.isMandatory() && loanCharge.isMandatoryInsurance())
-                        || (incident.isVoluntary() && loanCharge.isVoluntaryInsurance())) {
-                    BigDecimal cumulative = BigDecimal.ZERO;
-                    InsuranceIncidentNoveltyNews insuranceIncidentNoveltyNews = InsuranceIncidentNoveltyNews.instance(loan, loanCharge,
-                            null, incident, transactionDate, cumulative);
+    private void handleDefinitiveRestructuringCancellation(Loan loan, InsuranceIncident incident, LocalDate transactionDate) {
+        boolean doesIncidentExist = this.insuranceIncidentNoveltyNewsRepository.existsByLoanAndIncident(loan.getId(), incident.getId());
+        if (!doesIncidentExist) {
+            LoanCharge loanCharge = loan.getLoanCharges().stream()
```
Comment (faheem205) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:1398]: @oluexpert99 The purpose of below if condition is to make sure novelty news is created for either voluntary or mandatory charges or both depending upon the selection of charge types for the novelty news. Here you have restricted the code to create novelty news for only mandatory insurance even if the loan has voluntary insurance and voluntary insurance is selected for the novelty incident

**Review by faheem205 — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 1 above)
Comment (faheem205) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:1398]: @oluexpert99 existsByLoanAndIncident method should check if incident is created against a particular charge or not if needed

Comment (oluexpert99) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:1398]: Hi @faheem205  , This check is just for DEFINITIVE_RESTRUCTURING_CANCELLATION
The requirement is that the event be shown just once for loan .

Comment (faheem205) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:1398]: @oluexpert99 So if a loan has only voluntary insurance then this incident will not be recorded?

Comment (oluexpert99) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:1398]: @faheem205 ,Now I get what you are implying. Thanks for the catch . PR updated

Comment (faheem205) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:1398]: @oluexpert99 Now if loan does not have a mandatory or voluntary insurance charge then the code will try to create an incident with null loancharge and probably throw an exception

**Review by faheem205 — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 1 above)
Comment (oluexpert99) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:1398]: fixed

**Review by faheem205 — APPROVED** [review]

**Pull request merged**